Add support for user defined types

kumarlokesh opened the pull request:
Thank you for submitting this pull request. We really appreciate you spending the time to work on SurrealDB. 🚀 🎉 

## What is the motivation?

Attempts to resolve issue https://github.com/surrealdb/surrealdb/issues/5676

## What does this change do?

1. Allows defining custom types with a name and kind
2. Allows adding comments
3. Allows handling existence checks with `IF NOT EXISTS`
4. Extends the parser for the `DEFINE TYPE` statement
5. Performs type validation and checking when using custom types

## What is your testing strategy?

Added tests for the new functionality

## Is this related to any issues?

Fixes #5676 

## Does this change need documentation?

<!-- If this pull request requires changes, updates, or improvements to the documentation, then add a corresponding issue on the https://github.com/surrealdb/docs.surrealdb.com repository, and link to it here. -->

- [ ] No documentation needed

## Have you read the Contributing Guidelines?

<!-- All pull requests require that the contributing guidelines have been read and agreed to. -->

- [x] I have read the [Contributing Guidelines](https://github.com/surrealdb/surrealdb/blob/main/CONTRIBUTING.md)

kumarlokesh committed:
feature: add support for user defined types

LucyEgan commented:
Nice 🚀
What's the requirement for `PERMISSIONS` on a defined type? I would expect that anyone able to use a type will be a database or above user for it to be used in `DEFINE FIELD`, and from that point it should be the table/fields permissions that are used or does this have a 3rd stage of permission checking?

kumarlokesh committed:
do not specify additional permissions on a defined type

kumarlokesh commented:
> Nice 🚀 What's the requirement for `PERMISSIONS` on a defined type? I would expect that anyone able to use a type will be a database or above user for it to be used in `DEFINE FIELD`, and from that point it should be the table/fields permissions that are used or does this have a 3rd stage of permission checking?

@LucyEgan Thank you for the feedback.
Yes, there should be no requirement for specifying additional `PERMISSIONS` on a defined type.

Also, there's no third stage of permission checking that might be specific to types.
Once a type is defined, the table/field permissions should suffice.

I've updated the PR to remove references where permissions were associated with a type definition. Also, updated the PR description.

kearfy commented:
Hey @kumarlokesh, thank you for this amazing PR! We'd like to merge it in, but it requires a little bit more work. Would it be okay if I update it and push some further changes?

kumarlokesh commented:
> Hey @kumarlokesh, thank you for this amazing PR! We'd like to merge it in, but it requires a little bit more work. Would it be okay if I update it and push some further changes?

Sure @kearfy 👍

kearfy committed:
Merge remote-tracking branch 'origin/main' into pr/5729

kearfy committed:
WIP

kearfy commented:
@kumarlokesh to properly get this to work, all type coercions and convertions need to be async, which requires quite some work. I have other priorities at the moment, but I cannot continue work on this at the moment. We may pick it up again at a later stage

geldot commented:
It'd be interesting for user defined types to bring along a check/constraint, maybe something like:

```surql
DEFINE TYPE email AS string ASSERT string::is_email($value);
```

Other things to consider are user defined methods/casts/operators on custom types. The async cast problem could be addressed if custom types are opaque, requiring the definition of explicit/implicit cast functions. A survey of use cases and custom types in other databases might help avoid introducing duff syntax.

For example, what would be the MVP to support elegant and optimal IPv4/CIDRv4 user defined types? The representation might be a uint32 with user defined casts to/from string.

Another apparent use case might be "Felt252" #5331, which the patch author describes as "252-bit number used in the StarkNet ecosystem". This is a feature basically nobody wants in core, but is obviously important to that user.

The new custom type support doesn't need to implement all of this, but my sense is that it's worth doing some use case bikeshedding in advance to avoid the cycle of removal/rewrite past features have gone through (like record references, futures etc.)
